### ISSUE TITLE
Enable forwarding warnings from subprocess to the caller

### ISSUE THREAD

**Issue by pradyunsg** [pypa/pyproject-hooks#157]
This would be a reasonable mechanism to surface deprecation/build-related warnings from underlying build-backends to the build-frontends, which can then surface them to users appropriately.

This would likely involve some way to capture and communicate the log messages back to the calling process; ideally with the ability to show the warnings without waiting for the end of the build-backend subprocess.

**Comment by pradyunsg** [comment]
To be clear, I'm thinking of `warnings.warn`-style warnings.

**Comment by pradyunsg** [comment]
FYI: I've started exploring implementing this and figuring out what this should look like.

**Comment by pradyunsg** [comment]
Hmm... I came here to close this now that #213 is merged but it looks like this might actually need a PEP or something to make things clearer?

@notatallshaw Any concerns if I revert #213 and we resurrect that discussion? I might have missed context from a more recent discussion, if there is one.

**Comment by notatallshaw** [comment]
> [@notatallshaw](https://github.com/notatallshaw) Any concerns if I revert [#213](https://github.com/pypa/pyproject-hooks/pull/213) and we resurrect that discussion? I might have missed context from a more recent discussion, if there is one.

The comment here suggested this would just need a spec clarification not a PEP: https://github.com/pypa/packaging-problems/issues/558#issuecomment-981541735

There was some recent discussion in this thread: https://discuss.python.org/t/how-can-build-backends-avoid-breaking-users-when-they-make-backwards-incompatible-changes/85847

My motivation here is to help build backends communicate with users to reduce the chance of unexpected failures, and I don't really have an opinion on the spec, I haven't read it in detail. But if you're not comfortable with it I am motivated to do whatever steps are required until you are.

**Comment by pfmoore** [comment]
> The comment here suggested this would just need a spec clarification not a PEP

So that means it needs (1) a dedicated thread on Discourse proposing the change and establishing community consensus that it's ok as a spec change, and (2) the actual spec change having been made.

I agree with @pradyunsg - let's revert until that has happened.

**Comment by notatallshaw** [comment]
> So that means it needs (1) a dedicated thread on Discourse proposing the change and establishing community consensus that it's ok as a spec change, and (2) the actual spec change having been made.

I was being optimistic that this had already been done since your discussion.

I started take a look at doing this in the spec page, but as best as I can tell PEP 517 has never been added to: https://packaging.python.org/en/latest/specifications/

So as a first step should I raise a PR that adds PEP 517 to the spec page and then another PR that updates the PEP to link to the spec page (as I see with many other packaging PEPs)?

**Comment by pfmoore** [comment]
Wow, that was back in 2021? I hadn't even noticed the date 🙁

My general impression is that build backends don't tend to engage much with the standards process. That may be because PEP 517 is actually pretty good, and there's little need for change (optimistically, I hope that *is* the case). But it is surprising[^1] to me that setuptools identified this need back in 2021, but then didn't do anything to move it forward, leaving us in the situation now that they can't get their deprecation warnings out to users...

At some point, *someone* needs to get PEP 517 added to the specs, yes. If it's something you are willing to do, then go for it. Although the revision process - see below - does allow for updating specs where the PEP is still the reference documentation, so moving PEP 517 into the specs isn't a blocker here.

The actual process for making a change is [here](https://www.pypa.io/en/latest/specifications/#handling-fixes-and-other-minor-updates). I don't think what's being suggested here comes under the classification of "readability improvements that don’t affect software interoperability", unless we're talking about nothing more than a description of the fact that build backends *might* issue warnings, and frontends *might* choose to catch them. But once we move into the territory of suggesting that frontends should (or even must) display warnings, that's more than just readability.

At that point, the *actual* proposal should be raised on Discourse. That's not simply a broad statement of intent ("we should expose backend warnings"), it's a specific proposal for new wording, which will (for example) be explicit about whether backends can assume that warnings will be displayed, and what frontends have to do. The discussion can result in changes to the wording, or even completly new proposals, but we have to start with something specific. It's not a PEP as such[^2], but it *does* need to be similarly precise. The process isn't 100% clear on what is needed for a proposal to be "approved as a text-only change", but I've always assumed it requires community consensus, with the PEP-delegate (me) having final say that an appropriate consensus has been reached. If we don't get consensus, the proposal needs to go to a PEP or be dropped.

Personally, my view is that a change saying that frontends *should* provide a mechanism to allow users to see warnings issued by build backends would be acceptable. Saying that frontends which provide such a mechanism should display backend warnings by default would be more controversial, but might still be OK. But it would still mean that backends cannot rely on warnings being seen by default, and must be prepared for the possibility that warnings don't get shown (i.e., what we have now). Anything that *requires* front ends to display backend warnings feels like it's in PEP territory to me. But that's just my personal feel. Community consensus could be different - the only way to know is to make a proposal.

I will note that the mechanism here (as implemented in #171) is perfectly OK with no spec change. It's just that it counts as "implementation defined behaviour" currently. I don't know for certain what this project's position is on adding implementation-specific mechanisms, but I feel that as the reference implementation, it should be very cautious about doing so. Similarly, I know that pip's position (as a consumer) is generally that we don't add feaures unless they are backed by a standard. So the need for a spec change is more to do with project policies than because the current spec *disallows* this mechanism.

[^1]: And a little disappointing, if I'm honest.
[^2]: Although it can turn into one if necessary.

**Comment by pradyunsg** [comment]
Hmm... Thinking about this a bit more (and given the context of the more recent threads and Al that), I'm inclined to have us move forward here.

I'm happy to commit to updating PEP 517 spec / driving those discussions when we come around to adding it to the standards section on packaging.python.org.

AFAICT, no one has any blocking concerns with the actual logical change here and our concerns are mostly that we're deviating from the norms here / haven't done all the dotting the i-s, crossing the t-s.

**Comment by pradyunsg** [comment]
> At that point, the actual proposal should be raised on Discourse.

More than what was discussed in https://discuss.python.org/t/adding-a-mechanism-to-forward-user-facing-warnings-from-build-backends-to-build-frontends/28106?

**Comment by notatallshaw** [comment]
I would have liked to give a more detailed reply , but it would take me some time and I'm not likely to have that time soon.

So in short, I would not look to require front end installers to display warnings, my original thought was to only say that they MAY display warnings, and that they MAY make that choice on a per distribution basis. But I would also be happy to say that they SHOULD display warnings.

In a practical sense it is up to pyproject-hooks to forward on the warnings, and it would be up to pip to suppress them on a per distribution basis, based on whatever criteria pip ends up thinking is a good UX.

**Comment by pfmoore** [comment]
My only concerns are:

1. Are build backends OK with not being able to guarantee that warnings will be seen by users?
2. What is uv, in particular, going to do? They don't use this library, and they may have problems with accessing Python warnings.

The latter is only about the broader picture of how useful this will be in practice for backends.

Neither of these are blockers for adding this, either here or in pip.

**Comment by ofek** [comment]
* I strongly disagree that we should have custom logic for Python warnings; indeed as Paul said backends may not even be written in Python.
* I feel very strongly that we should modify the text to state that output streams should not be captured so that output from builders is shown in real time. That's all that is necessary.

**Comment by zahlman** [comment]
At the risk of repeating previous discussion, I think there are two key issues that need to get sorted out:

1. Like @ofek says, the backend isn't necessarily written in Python, and may be arbitrarily complex. If compilers are invoked, then those are separate processes again with their own IO streams, logging systems etc. However, there's still *some* Python involved, even using something like meson-python - i.e. the part that actually implements the PEP 517 hooks. So the question is, are we trying to prioritize warnings from the build *orchestration*, or are we trying to show the overall process? For example, if Setuptools verifies some config files, emits some warnings and then invokes a compiler, and we display everything, the compiler's output could scroll the initial warnings away. Are we okay with that?

2. As I see things, the real issue is that *developers* aren't seeing the warnings. But *users* also build wheels (implicitly via an installer) - possibly far more often than developers (certainly that's the case for some projects). Typically, though, users will find those warnings confusing and/or noise, and can't do anything about them except to report them. And then they have to decide whether to report it against the installed package or against the build backend. (Some might even try to report it against the installer....) Annoying users is *one* way to get developer attention, but probably not the best. Meanwhile, if developers use e.g. `pip wheel` as a build frontend (or maybe they hack together a worse idea...), they get the same experience as the end users - so hiding output is hiding output. And if they don't actually publish wheels, then they never directly see the project build. (Never mind the cases where the project is unmaintained; can't do anything about that.)

> My general impression is that build backends don't tend to engage much with the standards process. That may be because PEP 517 is actually pretty good, and there's little need for change (optimistically, I hope that is the case).

For this *specific* issue, I think a lot of it has to do with project age and the preferential PEP 518 treatment. Build backends other than Setuptools scarcely *have* deprecations to warn about, as far as I'm aware, don't have to think about projects that lack a `pyproject.toml` completely, etc.

But in general - I've seen other build backend developers in *discussion* a fair bit, but there aren't a whole lot of standards proposals that seem like they'd directly impact on build backends. At least, not until we get through the [meta-discussion](https://peps.python.org/pep-0777/) about the wheel standard. But the only really relevant standards are the ones that concern the content of sdists and the content of wheels. So, things like PEP 639, because that actually changes things on the back end.

**Comment by layday** [comment]
PEP 517 hooks must be implemented in Python regardless.  Backends must expose a Python interface and frontends must be able to execute Python.  The uv backend exposes Python hooks. The uv frontend runs Python to communicate with backends.  It does exactly what this library does - it calls the backend from a Python (sub)process and dumps its output in a JSON file.

It helps to say why you're feeling strongly about these things.

**Comment by notatallshaw** [comment]
@pfmoore 
> Are build backends OK with not being able to guarantee that warnings will be seen by users?

I'm not sure I follow the question, any user can send all stdout and stderr to dev null, there's never a way to guarantee the warning will be **seen**.

> What is uv, in particular, going to do? They don't use this library, and they may have problems with accessing Python warnings.

@charliermarsh can speak for himself, but I think I've been convincing that it is worth outputting the warnings, at least some of the time.


@ofek 
> backends may not even be written in Python

Even if the backend isn't pure Python it must have a Python component to be callable via PEP 517, so they can choose, or not choose, to build infrastructure to emit Python warnings. 

If we can't make requirements on a build backend having any ability to interact with Python then PEP 517 needs to be completely replaced.

> we should modify the text to state that output streams should not be captured so that output from builders is shown in real time. That's all that is necessary.

If users want to see the full build output they can already do that by enabling verbose flag in installers. It has been very strongly argued in the discourse threads that most users do not want to see any build output unless the is an error.

This is about having build backends having a specific channel to communicate with installers even when they have not opted in, and has been strongly requested from setuptools for a long time.

If you want to make an argument that frontend installers should never hide build backend output I would ask you start a separate thread on discourse, because I think others would have strong opinions on that.

**Comment by layday** [comment]
See also [this](https://github.com/pypa/packaging-problems/issues/558) discussion from 2021 where I suggested introducing a backend warning class (subclassing `UserWarning`) so that backends can selectively emit warnings for frontends and frontends can provide the option to raise backends warnings, à la pytest.

**Comment by ofek** [comment]
> If users want to see the full build output they can already do that by enabling verbose flag in installers. It has been very strongly argued in the discourse threads that most users do not want to see any build output unless the is an error.

Perhaps we can meet in the middle then and say that installers should expose user-configured verbosity to build backends. This way installers don't have to do anything special and the builders can use a standardized enum to influence their output. Like the person before mentioned, builds can be very complex and often call other processes such as when compiling extension modules which can take a long time and users often confuse lack of output with something being wrong/nothing happening.

I think this also solves the warning manifestation because builders can simply output them at the end of the build, or right before exiting with an exception.

**Comment by notatallshaw** [comment]
To quickly address some of the comments added while I was in the middle of my previous reply. 

Warning can affect users but not developers , and users may need to take action, such as applying build constraints. We have seen this with recent updates to setuptools and wheel that broke certain user environments, but not necessarily the developer of the package. Please see the discussion on discourse for more on this.

It will be up to the frontend installer (e.g. pip) to decide what warnings to suppress or not suppress based on providing a good user experience, not pyproject-hooks, so that discussion isn't relevant here.

@ofek 
> Perhaps we can meet in the middle then and say that installers should expose user-configured verbosity to build backends. This way installers don't have to do anything special and the builders can use a standardized enum to influence their output

This sounds like a feature request for pip, which should be filed against the pip issue tracker, or a new standard proposal on what installers SHOULD do, which should be created on discourse. It does not directly relate to this issue. 

These alternative proposals need agreement or consensus from pip maintainers or packaging community, and this is the wrong forum for that. And unless they achieve that agreement or consensus then all this discussion is doing is continuing to prolong the issue that backends have no way to communicate with users to using frontend installers. 

Can we please limit discussion to this issue and the merits of any objections against it, which I will try and spend some more time responding too in detail next week, but will just again remind participants that this is a specific request from the most popular build backend, and objections should take that into account.

**Comment by pfmoore** [comment]
This specific issue is about a very basic change to the interface presented by `pyproject-hooks`, to let frontends have access to Python warnings produced by the build backend. That's 100% compatible with the current spec, and as such, needs little discussion beyond "does it help the intended audience (i.e., the build backends who want to communicate with the end user)".

The other proposals, in particular anything needing a change to the spec, are big enough that they would need a discussion on the packaging Discourse, and probably a PEP. Can we please move any such conversations there, and leave this issue for discussing the specific `pyproject-hooks` change being proposed?

**Comment by zahlman** [comment]
I'm afraid I don't understand how it's compatible with the current spec, because I don't see how `pyproject-hooks` can *receive* the warnings from the backend in any selective way (i.e., other than by capturing all the process output) so as to forward them to the frontend.

**Comment by notatallshaw** [comment]
@zahlman by using Python's warnings mechanisms: https://docs.python.org/3/library/warnings.html

The relevant PR is already merged: https://github.com/pypa/pyproject-hooks/pull/213

**Comment by notatallshaw** [comment]
@layday I'll happily go over the code in more detail next week when I have more time, but quickly reviewing, no I don't think any warnings are suppressed or reordered.

The behavior collects and reemits warnings, I agree that we should check if/how this affects build and update the design as needed.

**Comment by layday** [comment]
I'm sorry, "rearrange" was ambiguous - what I meant was that it removed them from where they would've appeared originally and appends them at the bottom.  Warnings other than `UserWarning` aren't re-emitted either, from what I can tell.

**Comment by notatallshaw** [comment]
> what I meant was that it removed them from where they would've appeared originally and appends them at the bottom

I understood what you meant, I need to validate, but I don't think that's correct, I don't think any warnings is removed from their original location, and if it is correct I am in agreement it needs to be reworked.

I'm in the middle of travelling so I can't work on the code until next week.

**Comment by pradyunsg** [comment]
> Correct me if I'm wrong, but [#213](https://github.com/pypa/pyproject-hooks/pull/213) captures all warnings arising from the backend, preventing them from being printed to stdout.

That's... worth fixing. I think we should have that capture the warnings while still printing them as they would be printed without that capturing.

**Comment by hugovk** [comment]
CPython's C API deprecations marked with `__attribute__((deprecated))`  don't show up in the output when building with `pip install`, only in `pip install --verbose`:

```c
  src/_imagingft.c:145:13: warning: 'Py_FileSystemDefaultEncoding' is deprecated [-Wdeprecated-declarations]
    145 |             Py_FileSystemDefaultEncoding,
        |             ^
  /Library/Frameworks/Python.framework/Versions/3.13/include/python3.13/fileobject.h:22:1: note: 'Py_FileSystemDefaultEncoding' has been explicitly marked deprecated here
     22 | Py_DEPRECATED(3.12) PyAPI_DATA(const char *) Py_FileSystemDefaultEncoding;
        | ^
  /Library/Frameworks/Python.framework/Versions/3.13/include/python3.13/pyport.h:251:54: note: expanded from macro 'Py_DEPRECATED'
    251 | #define Py_DEPRECATED(VERSION_UNUSED) __attribute__((__deprecated__))
        |                                                      ^
  1 warning generated.

  src/_imagingft.c:145:13: warning: 'Py_FileSystemDefaultEncoding' is deprecated [-Wdeprecated-declarations]
    145 |             Py_FileSystemDefaultEncoding,
        |             ^
  /Library/Frameworks/Python.framework/Versions/3.13/include/python3.13/fileobject.h:22:1: note: 'Py_FileSystemDefaultEncoding' has been explicitly marked deprecated here
     22 | Py_DEPRECATED(3.12) PyAPI_DATA(const char *) Py_FileSystemDefaultEncoding;
        | ^
  /Library/Frameworks/Python.framework/Versions/3.13/include/python3.13/pyport.h:251:54: note: expanded from macro 'Py_DEPRECATED'
    251 | #define Py_DEPRECATED(VERSION_UNUSED) __attribute__((__deprecated__))
        |                                                      ^
```

This is unfortunate, as it means maintainers may miss deprecations and be surprised when they're finally removed.

It would be useful for pip to show these deprecations by default, or for an option we can recommend to maintainers (other than `--verbose`).

**Comment by notatallshaw** [comment]
> That's... worth fixing. I think we should have that capture the warnings while still printing them as they would be printed without that capturing.

Agreed, this should be the behavior if it is not already.

I've been meaning to validate this one way or the other, but my open source time in the last two weeks has been consumed by looking at resolver issues since the pip 25.1 release.
